docs/docsite/rst/dev_guide/developing_modules_best_practices.rst

```diff
@@ -1,197 +1,136 @@
+.. _developing_modules_best_practices:
 .. _module_dev_conventions:
 
+*****************************************
 Conventions, Best Practices, and Pitfalls
-`````````````````````````````````````````
+*****************************************
 
-As a reminder from the example code above, here are some basic conventions
-and guidelines:
+.. contents:: Topics
+   :local:
 
-* If the module is addressing an object, the parameter for that object should be called 'name' whenever possible, or accept 'name' as an alias.
+As you develop your module, follow these basic conventions and best practices:
 
-* If you have a company module that returns facts specific to your installations, a good name for this module is `site_facts`.
+General best practices
+======================
 
-* Modules accepting boolean status should generally accept 'yes', 'no', 'true', 'false', or anything else a user may likely throw at them.  The AnsibleModule common code supports this with "type='bool'".
+* Each module should be self-contained in one file, so it can be be auto-transferred by Ansible.
+* Always use the ``hacking/test-module`` script when developing modules - it will warn you about common pitfalls.
+* Validate upfront--fail fast and return useful and clear error messages.
+* Use defensive programming--use a simple design for your module, handle errors gracefully, and avoid direct stacktraces.
+* Fail predictably--if we must fail, do it in a way that is the most expected. Either mimic the underlying tool or the general way the system works.
+* If you have a local module that returns facts specific to your installations, a good name for this module is ``site_facts``.
+* If your module is addressing an object, the parameter for that object should be called ``name`` whenever possible, or accept ``name`` as an alias.
+* Modules accepting boolean status should generally accept ``yes``, ``no``, ``true``, ``false``, or anything else a user may likely throw at them. The AnsibleModule common code supports this with ``type='bool'``.
+* Eliminate or minimize dependencies. If your module has dependencies, document them at the top of the module file and raise JSON error messages when dependency import fails.
+* If you package your module(s) in an RPM, install the modules on the control machine in ``/usr/share/ansible``. Packaging modules in RPMs is optional.
+* Don't write to files directly; use a temporary file and then use the ``atomic_move`` function from ``ansible.module_utils.basic`` to move the updated temporary file into place. This prevents data corruption and ensures that the correct context for the file is kept.
+* Avoid creating caches. Ansible is designed without a central server or authority, so you cannot guarantee it will not run with different permissions, options or locations. If you need a central authority, have it on top of Ansible (for example, using bastion/cm/ci server or tower); do not try to build it into modules.
+* Handle exceptions (bugs) gracefully
```

Comment:
Hmmm. Is there a useful distinction between "module failures" (the module can't complete its task because, say, the permissions on a directory are incorrect) and "exceptions" (the code malfunctions because of a bug)?